crates/biome_js_type_info/src/flattening/expressions.rs

```diff
@@ -61,132 +61,149 @@ pub(super) fn flattened_expression(
-                            .to_data()
-                            .find_type_of_elements_from_index(
-                                resolved.resolver_id(),
-                                resolver,
-                                *index,
-                            )
-                            .map_or_else(TypeData::unknown, ResolvedTypeData::to_data),
-                    ),
-                    (TypeData::InstanceOf(subject_instance), DestructureField::Name(name)) => {
-                        resolver
-                            .resolve_and_get(
-                                &resolved.apply_module_id_to_reference(&subject_instance.ty),
-                            )
-                            .and_then(|subject| {
-                                subject.all_members(resolver).find(|member| {
-                                    !member.is_static() && member.has_name(name.text())
+            let resolved = resolver.resolve_and_get(&expr.ty)?;
+            match (resolved.as_raw_data(), &expr.destructure_field) {
+                (_subject, DestructureField::Index(index)) => Some(
+                    resolved
+                        .to_data()
+                        .find_element_type_at_index(resolved.resolver_id(), resolver, *index)
+                        .map_or_else(TypeData::unknown, ResolvedTypeData::to_data),
+                ),
+                (_subject, DestructureField::RestFrom(index)) => Some(
+                    resolved
+                        .to_data()
+                        .find_type_of_elements_from_index(resolved.resolver_id(), resolver, *index)
+                        .map_or_else(TypeData::unknown, ResolvedTypeData::to_data),
+                ),
+                (TypeData::InstanceOf(subject_instance), DestructureField::Name(name)) => resolver
+                    .resolve_and_get(&resolved.apply_module_id_to_reference(&subject_instance.ty))
+                    .and_then(|subject| {
+                        subject
+                            .all_members(resolver)
+                            .find(|member| !member.is_static() && member.has_name(name.text()))
+                    })
+                    .and_then(|member| resolver.resolve_and_get(&member.deref_ty(resolver)))
+                    .map(ResolvedTypeData::to_data),
+                (TypeData::InstanceOf(subject_instance), DestructureField::RestExcept(names)) => {
+                    resolver
+                        .resolve_and_get(
+                            &resolved.apply_module_id_to_reference(&subject_instance.ty),
+                        )
+                        .map(|subject| {
+                            // We need to look up the prototype chain, which may
+                            // yield duplicate member names. We deduplicate
+                            // using a map before constructing a new object.
+                            let members: BTreeMap<Text, ResolvedTypeMember> = subject
```

Comment:
Not relevant to this PR, but why do we order members based on their name, instead - for example - text range?

Reply:
Thank you. Would you mind a comment so we don't forget? Also, does the `HashMap` increase performance? Asking because last time this helped

Reply:
Will do, and I'll give it a shot with a hash table indeed :)